drivers/python/rethinkdb/net_asyncio.py

```diff
@@ -0,0 +1,236 @@
+# Copyright 2015 RethinkDB, all rights reserved.
+
+import asyncio
+import contextlib
+import socket
+import struct
+
+from . import ql2_pb2 as p
+from .net import decodeUTF, Query, Response, Cursor, maybe_profile, convert_pseudo
+from .net import Connection as ConnectionBase
+from .errors import *
+
+__all__ = ['Connection']
+
+pResponse = p.Response.ResponseType
+pQuery = p.Query.QueryType
+
+@asyncio.coroutine
+def _read_until(streamreader, delimiter):
+    """Naive implementation of reading until a delimiter"""
+    buffer = bytearray()
+    while True:
+        c = yield from streamreader.read(1)
+        if c == b'':
+            break  # EOF
+        buffer.append(c[0])
+        if c == delimiter:
+            break
+    return bytes(buffer)
+
+@contextlib.contextmanager
+def translate_timeout_errors():
+    try:
+        yield
+    except asyncio.TimeoutError:
+        raise RqlTimeoutError
+
+
+# The asyncio implementation of the Cursor object:
+# The `new_response` Future notifies any waiting coroutines that the can attempt
+# to grab the next result.  In addition, the waiting coroutine will schedule a
+# timeout at the given deadline (if provided), at which point the future will be
+# errored.
+class AsyncioCursor(Cursor):
+    def __init__(self, *args, **kwargs):
+        Cursor.__init__(self, *args, **kwargs)
+        self.new_response = asyncio.Future()
+
+    def _extend(self, res):
+        Cursor._extend(self, res)
+        self.new_response.set_result(True)
+        self.new_response = asyncio.Future()
+
+    # Convenience function so users know when they've hit the end of the cursor
+    # without having to catch an exception
+    @asyncio.coroutine
+    def fetch_next(self, wait=True):
+        timeout = Cursor._wait_to_timeout(wait)
+        while len(self.items) == 0 and self.error is None:
+            self._maybe_fetch_batch()
+            with translate_timeout_errors():
+                yield from asyncio.wait_for(self.new_response, timeout)
```

Comment:
The Tornado implementation uses a deadline rather than a timeout because if this loop is performed multiple times, we don't want to exceed the original timeout - perhaps implementing an analog to `with_absolute_timeout` from `net_tornado.py` would help.